openghg/objectstore/metastore/_metastore.py

```diff
@@ -0,0 +1,242 @@
+"""
+This module defines an interface for metastores.
+
+Metastores store metadata that is used to search an
+object store for data.
+
+The documentation in this module refers to metadata entries stored
+in the metastore as "records". These correspond to "documents" in
+TinyDB and Mongodb.
+
+A MetaStore managed by an ObjectStore object will store
+UUIDs, which are used to locate data stored in the object store.
+
+UUIDs are managed by ObjectStore objects, so they are not part of
+the MetaStore interface.
+
+"""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Optional, Union
+
+import tinydb
+from tinydb.operations import delete as tinydb_delete
+
+from openghg.types import MetastoreError
+
+MetaData = Dict[str, Any]
+QueryResults = List[Dict[str, Any]]  # ...to avoid clashes with `SearchResults` object
+Bucket = str
+
+
+class MetaStore(ABC):
+    """Interface for MetaStore.
+
+    All classes implementing this interface must provide the following methods:
+        search: to search for records in the metastore matching given key-value pairs
+        add: to add metadata to a new record in the metastore
+        delete: to delete records matching the given key-value pairs
+        update: update the metadata of a record matching the `where` key-value pairs;
+            keys may be added or updated (using `to_update`) or deleted (using `to_delete`)
+    """
+
+    @abstractmethod
+    def search(self, search_terms: Optional[MetaData] = None) -> QueryResults:
+        """Search for data using a dictionary of search terms.
+
+        Args:
+            search_terms: dictionary of key-value pairs to search by.
+
```

Comment:
I guess the google style is to have empty lines before Args, Returns, and Raises: https://google.github.io/styleguide/pyguide.html#doc-function-args (also, no type hints in the docstring if type annotations are used, and no return if first line is sufficient, or return type is None)

We don't need to have empty lines though (or do the other google docstring stuff). I guess I'm doing the "raises" part wrong according to the link.

Maybe in the developer docs we can write out what we want, either google style verbatim, or google style with X changes (or just write out examples and just say "similar" to google style).

(Also, I haven't got my text editor to auto insert docstrings yet... I used to have it. Does VSCode automatically do the style you prefer?)

Reply:
I think in general it doesn't matter which style we adopt as long as we're clear and we do this consistently.

We should make sure that it's clear what style we want to use and that everyone is on board. In the current documentation I can see we have this (put together by Gareth): https://github.com/openghg/openghg/blob/devel/doc/source/development/python_devel.rst#coding-style (https://github.com/openghg/openghg/blob/devel/doc/source/contributing.rst#contributing-new-code has a broken link to presumably this page as well?). This includes some examples of doc-strings but not necessarily in the style being applied here.

My suggestion would be that, if we want to apply a style as has been requested here this should be in this style guide. That way this can be referenced (and agreed) when having these discussions.

Reply:
To that end I've opened up an issue - https://github.com/openghg/openghg/issues/782. This may seem trivial but I think it's important we get this agreed so we can help us easily keep a consistent style (less review in PRs as well) - @gareth-j @brendan-m-murphy Can one of you take this on and then we can finalise this? Hopefully shouldn't take too long to just get this written down.

Reply:
Sorry I forgot that the was documentation about this already